Modern OXS Specification (#1)
embroidery-space/xsp-specs: ursa/oxs.md

> # OXS (Open Cross-Stitch)

Author: ```suggestion
# Open Cross-Stitch File Format (OXS)
```

OXS is an open file format for cross-stitch patterns that was created by [UrsaSoftware](https://ursasoftware.com).

Since OXS is built on XML, it is simple to examine its structure.
The format is highly extensible, allowing users to add their own attributes and define custom sections and elements.

This document offers a detailed specification of the OXS format, extending the original version.
It adds thorough explanation of each section and element, outlines the usage of each attribute and defines the expected behavior of parsers.
There's also a collection of application-specific items and additional examples.

> You can find a practical OXS parser [here](https://github.com/embroidery-space/embroidery-studio/blob/main/src-tauri/src/core/parsers/oxs.rs).

Author: ```suggestion
> You can find a practical OXS parser [here](https://github.com/embroidery-space/embroidery-studio/blob/main/src-tauri/src/core/parsers/oxs.rs).
```

# OXS (Open Cross-Stitch)

OXS is an open file format for cross-stitch patterns that was created by [UrsaSoftware](https://ursasoftware.com).

Since OXS is built on XML, it is simple to examine its structure.
The format is highly extensible, allowing users to add their own attributes and define custom sections and elements.

This document offers a detailed specification of the OXS format, extending the original version.

> It adds thorough explanation of each section and element, outlines of the usage of each property and describes its expected behavior.

Author: ```suggestion
It adds thorough explanation of each section and element, outlines the usage of each attribute and defines the expected behavior of parsers.
```

# OXS (Open Cross-Stitch)

OXS is an open file format for cross-stitch patterns that was created by [UrsaSoftware](https://ursasoftware.com).

Since OXS is built on XML, it is simple to examine its structure.
The format is highly extensible, allowing users to add their own attributes and define custom sections and elements.

This document offers a detailed specification of the OXS format, extending the original version.
It adds thorough explanation of each section and element, outlines the usage of each attribute and defines the expected behavior of parsers.
There's also a collection of application-specific items and additional examples.

> You can find a practical OXS parser [here](https://github.com/embroidery-space/embroiderly/blob/main/src-tauri/src/core/parsers/oxs.rs).

Author: ```suggestion
> You can find a practical OXS parser [here](https://github.com/embroidery-space/embroiderly/blob/main/src-tauri/src/core/parsers/oxs.rs).
```

# Open Cross-Stitch File Format (OXS)

OXS is an open file format for cross-stitch patterns that was created by [UrsaSoftware](https://ursasoftware.com).

Since OXS is built on XML, it is simple to examine its structure.
The format is highly extensible, allowing users to add their own attributes and define custom sections and elements.

This document offers a detailed specification of the OXS format, extending the original version.
It adds thorough explanation of each section and element, outlines the usage of each attribute and defines the expected behavior of parsers.
There's also a collection of application-specific items and additional examples.

> You can find a practical OXS parser [here](https://github.com/embroidery-space/embroiderly/blob/main/src-tauri/src/core/parsers/oxs.rs).

## Conventions

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

> ### Specification Conventions

Author: TODO: specify that application-specific elements are signed as `_by <software or company>_`.
List all the applications whose specific elements we are specifying. This will mostly be used by ourselves so that we don't make a mistake in the name of the application/company we use.

This specification defines both generic and application-specific elements. It is assumed that the common elements should be supported by all applications. However, this specification does not define _absolutely_ all of the specific elements.

  comments03="The features of each software package varies, but using XML each can pick out the things it can deal with, while ignoring others"
  comments04="The basic items are :"
  comments05="'palette'..a set of colors used in the design: palettecount excludes cloth color, which is item 0"
  comments06="'fullstitches'.. simple crosses"
  comments07="'backstitches'.. lines/objects with a start and end point"
  comments08="(There is a wide variety of ways of treating part stitches, knots, beads and so on.)"
  comments09="Colors are expressed in hex RGB format."
  comments10="Decimal numbers use US/UK format where '.' is the indicator - eg 0.5 is 'half'"
  comments11="For readability, please use words not enumerations"
  comments12="The properties, fullstitches, and backstitches elements should be considered mandatory, even if empty"
  comments13="element and attribute names are always lowercase"
/>
```

### `properties`

This section is OPTIONAL and defines general pattern properties.
It is RECOMMENDED to keep this section at the top of the file, as it can be used as a reference for parsing the pattern file.

Attributes:

- `oxsversion`: string - The version of the OXS format.

>   If not specified, we assume that the latest OXS version is used (currently, `1.0`).

Reviewer: Then don't specify the last part if it's unknown

>   If not specified, we assume that the latest OXS version is used (currently, `1.0`).

Author: But we must assume a specific OXS version. Probably, not `the latest` but just 1.0.

>   If not specified, we assume that the latest OXS version is used (currently, `1.0`).

Reviewer: Rephrase then

  Can be used to indicate app-specific elements and attributes.
- `software_version`: string - Specifies the software version used to write the pattern file.
  Can be used to indicate app/version-specific elements and attributes.
- `chartheight`: integer - Defaults to `100`.
- `chartwidth`: integer - Defaults to `100`.
- `charttitle`: string - Defaults to the name of the file.
- `author`: string.
- `copyright`: string.
- `instructions`: string.
- `stitchesperinch`: number - Specifies the number of stitches per inch of cloth/fabric.
- `stitchesperinch_y`: number - Same as `stitchesperinch`, but for vertical axis.
  Can be used for non-square cloth/fabric.
  If not specified, the value from `stitchesperinch` can be used.
- `palettecount`: integer - The number of palette colors other than cloth/fabric.

All of these attributes are OPTIONAL.
However, it is RECOMMENDED that you specify at least `oxsversion` for consistency, and `chartwidth` and `chartheight` for correct display.

If some of the attributes are not specified but are required by a particular software, they can be replaced by the default value of that software.

> 

Author: What do you mean? Didn't we already provide code blocks with examples?

> 

Reviewer: I mean it should have a word like "example" near (ideally above) them

> 

Author: ```suggestion

Example:

```

- `comments`: string.

Cloth/fabric-specific attributes:

- `kind` _by Embroidery_: string - Specifies the kind of the cloth/fabric.
  The common values are "Aida", "Evenweave", "Linen".

Material-specific attributes (threads, beads, etc.):

- `number`: string - Specifies both the color brand and the color number (for example, `DMC 310`).

  In Ursa, they are separated by four spaces.
  In other programs, they can be separated by a single space.

  For some reason, WinStitch adds `[+]` to the end of the `number` if the color is a blend.
  Therefore, it is RECOMMENDED to remove it to correctly parse the color number.

  In a normalized string, the last part of the string splitted by a space is the number (we expect the number to be a solid string), and everything before it is the brand.

- `printcolor`: rgb - Specifies the color used for color printing on paper.
- `blendcolor`: rgb - Specifies the color blended with the base color.
- `strands`: integer - The number of thread for stitching.
  If specified, it SHOULD be in the range of 1 to 6.
- `symbol`: integer or string - Specifies the symbol used to graphically represent the color.

>   It can be a decimal number representing a UTF-8 [code point](https://developer.mozilla.org/en-US/docs/Glossary/Code_point) or a string representing the actual character.

Author: ```suggestion
  It can be a number representing a UTF-8 [code point](https://developer.mozilla.org/en-US/docs/Glossary/Code_point) or a string representing the actual character.
```

- `fontname` _by XSPro Platinum (DP Software)_: string - Specifies the font family (for example, `Cross Stitch Pro Platinum`) used to draw symbols of this color.
- `metalic` _by XSPro Platinum (DP Software)_: boolean - Specifies whether the color represents a metalic thread.
- `fluorescent` _by XSPro Platinum (DP Software)_: boolean - Specifies whether the color represents a fluorescent thread.
- `colorcmyk`, `bscolorcmyk`, `printcolorcmyk` _by XSPro Platinum (DP Software)_: cmyk - Specifies the color of the corresponding attribute in the CMYK color model.
  These are important for publishers who will likely print their patterns on paper and want to achieve high repeatability thread color reproduction.

All of these attributes except `color` are OPTIONAL.
If the `color` attribute is missing, empty or `nil`, the application SHOULD replace it with the default color (see notes above) and notify the user about it so that they can fix the issue later.

The `palette_item` element with index `0` (usually, the first one) MUST always define the cloth/fabric color.
If the `index` attribute is not specified, the first `palette_item` in the palette is assumed to define the cloth/fabric color.

> 

Author: ```suggestion

Example:

```

```xml
<palette>
  <!-- The element with index 0 defines the cloth/fabric color. -->
  <palette_item index="0" number="cloth" name="cloth" color="FFFFFF" />

  <!-- Other elements define the material colors. -->
  <palette_item index="1" number="DMC    310" name="Black" color="2C3225" printcolor="000000" strands="2" symbol="100" bsstrands="1" bscolor="2C3225" />
  <palette_item index="2" number="DMC 158 [+]" name="DMC 158 [+] DMC 208" color="303065" blendcolor="824596" printcolor="593A7D" />
</palette>
```

##### `blend`

This is a special element used by Embroiderly to store additional information about blends.

Using the `blendcolor` attribute of the `palette_item` element you can specify only one color which is blended with the main color.
In contrast, using the `blend` element you can specify any number of colors used in a blend.

Attributes:

> - `number`: string.

Author: Until we don't have any other attributes, it is fine to declare it for the whole block. When we add other specific attributes, it will make sense to declare this for each element instead.

> - `number`: string.

Reviewer: Mate, it's not fine if it's:
1. non consistent
2. confusing

> - `number`: string.
> - `name`: string.
> - `color`: rgb.
> - `strands`: integer.
> 
> These attributes are the same as in the `palette_item` element.

Author: ```suggestion
- `number`: string - The same as in the `palette_item` element.
- `name`: string - The same as in the `palette_item` element.
- `color`: rgb - The same as in the `palette_item` element.
- `strands`: integer - The same as in the `palette_item` element.
```

Reviewer: Can you just copy paste the explanation of what it does instead?

> 

Author: ```suggestion

Example:

```

The stitch is considered to be "invalid" and MUST NOT be processed, stored or displayed in such cases:

1. If any stitch coordinate is missing or invalid (i.e., it cannot be converted from a string type to a numeric type).

   However, coordinates that are outside the pattern are considered correct.
   For example, if the size of the pattern is not specified in the `properties` section, some stitches may appear outside the pattern.
   We SHOULD process and store such stitches and let the user process them (for example, increase the size of the pattern or remove extra stitches).

2. If the `palindex` attribute is set to `0` (i.e., the stitch refers to a cloth/fabric color).

   Obviously, you SHOULD NOT store such stitches in the pattern file.

3. If the stitch refers to a non-existent palette item.
4. If the stitch is supposed to have an `objecttype` attribute, but it is missing or empty.

### `fullstitches`

Holds `stitch` elements.

If the `fullstitches` tag is missing or empty, this means that the pattern does not contain full stitches.

#### `stitch`

Defines full stitch information.

Attributes:

- `x`: number.
- `y`: number.
- `palindex`: integer.
- `marked`: boolean.

> <figure>

Author: It would be really nice, but in this case, the XML example (as well as the backstitches example) is too obvious.
The only critical case is partstitches. And there, the XML example is exactly aligned with the image.

> <figure>

Reviewer: No, it's really not
It is confusing to a person who are trying to make sense of it, not the person who wrote it

  <img src="../images/ursa/oxs/fullstitches.png" alt="Full stitches in WinStitch."/>
  <figcaption>
    Full stitches in WinStitch.
  </figcaption>
</figure>

> 

Author: ```suggestion

Example:

```

```xml
<fullstitches>
  <!-- This stitch is "empty" because it uses the cloth/fabric color. -->
  <stitch x="1" y="1" palindex="0" />

  <!-- These are normal stitches. -->
  <stitch x="2" y="2" palindex="1" />
  <stitch x="3" y="3" palindex="2" />
  <stitch x="4" y="4" palindex="3" marked="true" />
</fullstitches>
```

### `partstitches`

Holds `partstitch` elements.

#### `partstitch`

Defines part stitch information.

Attributes:

- `x`: number.
- `y`: number.
- `palindex1`: integer.

  For three-quarter stitches, specifies the color on the left.

  For half/gobelin stitches, specifies the color of the entire stitch.

- `palindex2`: integer.

  Only applicable for three-quarter stitches.
  Specifies the color on the right.

- `direction`: integer - Specifies the direction of the part stitch.

  The possible values are:

  - `1` and `2` for three-quarter stitches.
  - `3` and `4` for half/gobelin stitches.

- `marked`: boolean.

> <figure>

Author: The numbers on the image are grouping stitches. Then, in the description, there is an ordered list aligned with those number groups.

> <figure>

Reviewer: I know what numbers represent
But it's not what happens when you see the 1. 2. 3. 4. for the first time in this block and then image has a very different meaning for those numbers
This is unacceptable to leave in this state

  <img src="../images/ursa/oxs/partstitches.png" alt="Part stitches in WinStitch."/>
  <figcaption>
    Half/gobeling and three-quarter stitches in WinStitch.
  </figcaption>
</figure>

1. Half/Gobelin stitches:

   - The top one is with `direction=3` (forward).
   - The bottom one is with `direction=4` (backward).

2. Three-quarter stitches:

   - The top-left one is with `direction=2`, `palindex1` set and `palindex2` not set.
   - The top-right one is with `direction=1`, `palindex1` not set and `palindex2` set.
   - The bottom-left one is with `direction=1`, `palindex1` set and `palindex2` not set.
   - The bottom-right one is with `direction=2`, `palindex1` not set and `palindex2` set.

Three-quarter stitches usually represent a single object (and are therefore drawn using a single stitch tool).
However, some applications MAY represent them as a combination of a quarter (1/2 of a half/gobelin stitch) or petit (1/4 of a full stitch) stitch and a half/gobelin stitch.

A single three-quarter stitch can have colors for both sides, so it will fill the entire cell.
Note that there cannot be more than one `partstitch` object in a cell, although this is not a significant problem.

> 

Author: ```suggestion

Example:

```

> Note that there cannot be more than one `partstitch` object in a cell, although this is not a significant problem.

Author: "Not a significant problem" means that one stitch may overlap other stitches. In such a case, they will be rendered without any issues.

> Note that there cannot be more than one `partstitch` object in a cell, although this is not a significant problem.

Reviewer: You should use explicit language, as you defined at the start of the document

> Note that there cannot be more than one `partstitch` object in a cell, although this is not a significant problem.

Author: ```suggestion
```

  - `bugle` _by Ursa Software_ - A stitch with a bugle bead.

    <figure>
      <img src="../images/ursa/oxs/backstitches-bugle.png" alt="Bugle beads in WinStitch."/>
      <figcaption>
        Bugle beads in WinStitch.
      </figcaption>
    </figure>

    <figure>
      <img src="../images/ursa/oxs/backstitches-bugle-real.png" alt="Bugle beads in reality."/>
      <figcaption>
        Bugle beads in reality.
      </figcaption>
    </figure>

  - `straightstitch` _by Embroiderly_ - A long back stitch.

    In Embroiderly, back stitches can be only one cell long.
    Straight stitches, in contrast, can be any length.

  - `curvedstitch` _by Embroiderly_ - A curved stitch.

    In Embroiderly, curved stitches are part of special stitch models.

    Curved stitches have an unspecified number of `x` and `y` coordinates.

  The software that does not actually support `daisy`, `bugle` or `straightstitch` objects MAY process them as `backstitch`.

- `sequence`: integer - Specifies the order in which the elements are rendered.
  In fact, it represents the z-index.
  Objects with a higher value of `sequence` overlap objects with a lower value of `sequence`.

- `marked`: boolean.

> 

Author: ```suggestion

Example:

```

```xml
<backstitches>
  <backstitch x1="2" x2="5" y1="2" y2="2.5" palindex="3" objecttype="backstitch" sequence="0"/>
  <backstitch x1="3" x2="4" y1="3" y2="2" palindex="1" objecttype="daisy" sequence="0"/>
  <backstitch x1="6" x2="6" y1="2" y2="5" palindex="1" objecttype="bugle" sequence="0"/>
  <backstitch
    x1="1.00" y1="0.00"
    x2="0.43" y2="0.26"
    x3="0.06" y3="0.66"
    x4="0.06" y4="1.03"
    x5="0.50" y5="1.06"
    x6="0.89" y6="0.66"
    x7="1.10" y7="0.10"
    palindex="1" objecttype="curvedstitch"
  />
</backstitches>
```

### `ornaments_inc_knots_and_beads`

Holds `object` elements.

If the `ornaments_inc_knots_and_beads` tag is missing or empty, this means that the pattern does not contain additional objects.

#### `object`

> The `object` element defines anything that can be added by other software.

Author: What is the problem here?

> The `object` element defines anything that can be added by other software.

Reviewer: What is non-other software then?

> The `object` element defines anything that can be added by other software.

Author: ```suggestion
The `object` element defines anything that can be added by a specific software.
```

      If not specified or `true`, we consider this stitch to be a **petit** stitch.
      Otherwise, we consider this stitch to be a **quarter** stitch.

  - `tent` _by XSPro Platinum (DP Software)_ - A half stitch.
    It MAY represent the type of stitches as the `partstitch` elements with the `direction` set to `3` or `4`.

    Additional attributes:

    - `direction`: integer - Specifies the direction of the part stitch.

      The possible values are:

      - `1` - backward.
      - `2` - forward.

  - `horizontalhalf`, `verticalhalf` _by UrsaSoftware_ - A stitch with half size of a full stitch.

    <figure>
      <img src="../images/ursa/oxs/hv-halfstitches.png" alt="Horizontal and vertical half stitches in WinStitch."/>
      <figcaption>
        Horizontal and vertical half stitches in WinStitch.
      </figcaption>
    </figure>

  - `fullcross` _by UrsaSoftware_ - A normal cross stitch which, unlike a `fullstitch`, can be placed anywhere on the cloth/fabric.

    <figure>
      <img src="../images/ursa/oxs/fullcrosses.png" alt="Full croses in WinStitch."/>
      <figcaption>
        Full croses in WinStitch.
      </figcaption>
    </figure>

>   - `3x2`, `2x3`, `3x3`, `4x4`, `4x2`, `6x2`, `8x2`, `10x2`, `12x2`, `16x2` _by UrsaSoftware_ - Normal stitches that are slightly larger than usual.

Reviewer: Slightly? (:

Author: Some slightly, some even more. Anyway, they are bigger than other stitches, aren't they?

Reviewer: Yes, but slightly does not apply for most listed
So maybe don't use that word? 😉

Author: ```suggestion
  - `3x2`, `2x3`, `3x3`, `4x4`, `4x2`, `6x2`, `8x2`, `10x2`, `12x2`, `16x2` _by UrsaSoftware_ - Normal stitches that are larger than usual.
```

    They are typically used for knitting.

    <figure>
      <img src="../images/ursa/oxs/large-stitches.png" alt="Large stitches in WinStitch."/>
      <figcaption>
        Large stitches in WinStitch.
      </figcaption>
    </figure>

  - `topleftlongtriangle` `botrightlongtriangle` `toprightlongtriangle` `botleftlongtriangle` `topleftwidetriangle` `botrightwidetriangle` `botleftwidetriangle` `toprightwidetriangle` _by UrsaSoftware_ - Triangle stitches.

    <figure>
      <img src="../images/ursa/oxs/triangle-stitches.png" alt="Triangle stitches in WinStitch."/>
      <figcaption>
        Triangle stitches in WinStitch.
      </figcaption>
    </figure>

  - `knot` - A french knot.

    <figure>
      <img src="../images/ursa/oxs/french-knots.png" alt="French knots in WinStitch."/>
      <figcaption>
        French knots in WinStitch.
      </figcaption>
    </figure>

  - `bead` - A bead.

>     In WinStitch, it is 2 mm in size.
> 
>     Additional attributes:
> 
>     - `length`, `diameter` _by Embroiderly_: number - Bead dimensions.
> 

Author: ```suggestion
    In WinStitch, it is 2 mm in size.

```

Author: Since we have not achieved a consensus on storing the bead data, it is better to remove these additional attributes.

  - `bead1mm`, `bead2.5mm`, `bead3mm`, `bead5mm`, `bead6mm`, `bead8mm`, `bead12mm` _by UrsaSoftware_ - Beads with the size specified.

    <figure>
      <img src="../images/ursa/oxs/beads.png" alt="Beads in WinStitch."/>
      <figcaption>
        Beads in WinStitch.
      </figcaption>
    </figure>

  - `specialstitch` _by Embroiderly_ - A special stitch from the XSD pattern.

    Additional attributes:

    - `modindex`: integer - Specifies the index of the special stitch model defined in the `special_stitch_models` section.
      If it is missing or empty, we consider this special stitch to be invalid.
    - `rotation`: integer - Specifies the angle of rotation of the stitch in degrees.
    - `flip_x`, `flip_y`: boolean - Specifies whether the stitch should be flipped along the corresponding axis.

- `marked`: boolean.

> 

Author: ```suggestion

Example:

```

  <object x1="3" y1="1" palindex="2" objecttype="toprightlongtriangle"/>
  <object x1="3" y1="2" palindex="2" objecttype="botleftlongtriangle"/>
  <object x1="1" y1="5" palindex="3" objecttype="topleftwidetriangle"/>
  <object x1="2" y1="5" palindex="3" objecttype="botrightwidetriangle"/>
  <object x1="1" y1="7" palindex="5" objecttype="botleftwidetriangle"/>
  <object x1="2" y1="7" palindex="5" objecttype="toprightwidetriangle"/>

  <!-- French knot. -->
  <object x1="1" y1="1" palindex="2" objecttype="knot" />

  <!-- Beads. -->
  <object x1="1.5" y1="1.5" palindex="1" objecttype="bead1mm" />
  <object x1="3.5" y1="1.5" palindex="1" objecttype="bead" />
  <object x1="5.5" y1="1.5" palindex="1" objecttype="bead2.5mm" />

  <!-- A special stitch. -->
  <object x1="10" y1="5.5" palindex="3" modindex="0" rotation="90" flip_x="true" flip_y="false" objecttype="specialstitch"/>
</ornaments_inc_knots_and_beads>
```

### `special_stitch_models` _by Embroiderly_

Holds `model` elements.

If the `special_stitch_models` tag is missing or empty, this means that the pattern does not contain special stitch models.

#### `model`

> Defines special stitch model information.

Reviewer: Where is the modindex? How do I address the special stitch I want?
It doesn't make sense to have the be indexed by only position in object, this is easily changeable and will create problems

Author: You read the special stitches models section (as in the case of the palette) and use the modindex to find the special stitch model.

Reviewer: Again, how would I know modindex? Why don't specify it explicitly (like palindex for example)

- `unique_name`: string.
- `name`: string.
- `width`: number.
- `height`: number.

The `model` element contains `backstitch` (with object types `backstitch`, `straightstitch`, `curvedstitch`) and `object` elements (with object type `knot`).

The `palindex` attribute is not actually used for special stitch models, but is specified to make their elements valid.

> 

Author: ```suggestion

Example:

```

> The `palindex` attribute is not actually used for special stitch models, but is specified to make their elements valid.

Reviewer: I'm not sure what this means

Author: The palindex is used only to make these elements valid OXS entities. But it is not used to render the special stitch model. Instead, when we use a single palindex from the object element, which has the specialstitch objecttype, to render the whole special stitch.

Reviewer: Still not entirely clear, needs more examples or rephrasing

Author: ```suggestion
The `palindex` attribute of the model's element is always `0`.
Thus, the model elements do not actually have a color.
Instead, when we render a special stitch that has this model, we use the color of that stitch for all model elements.
```

```xml

> <special_stitch_models>
>   <model unique_name="Rhodes Heart - over 6" name="Rhodes Heart" width="3.0" height="2.5">
>     <backstitch x1="1.0" x2="2.0" y1="2.0" y2="0.0" palindex="1" objecttype="straightstitch"/>
>     <backstitch x1="0.5" x2="2.5" y1="1.5" y2="0.0" palindex="1" objecttype="straightstitch"/>
>     <backstitch x1="0.0" x2="3.0" y1="1.0" y2="0.5" palindex="1" objecttype="straightstitch"/>
>     <backstitch x1="0.0" x2="3.0" y1="0.5" y2="1.0" palindex="1" objecttype="straightstitch"/>
>     <backstitch x1="0.5" x2="2.5" y1="0.0" y2="1.5" palindex="1" objecttype="straightstitch"/>
>     <backstitch x1="1.0" x2="2.0" y1="0.0" y2="2.0" palindex="1" objecttype="straightstitch"/>
>     <backstitch x1="1.5" x2="1.5" y1="0.5" y2="2.5" palindex="1" objecttype="straightstitch"/>
>   </model>
>   <model unique_name="Lazy Daisy" name="Lazy Daisy" width="1.0" height="1.5">
>     <backstitch
>       x1="1.00" y1="0.00"
>       x2="0.43" y2="0.26"
>       x3="0.06" y3="0.66"
>       x4="0.06" y4="1.03"
>       x5="0.50" y5="1.06"
>       x6="0.89" y6="0.66"
>       x7="1.10" y7="0.10"
>       palindex="1" objecttype="curvedstitch"
>     />
>     <backstitch
>       x1="0.03" y1="1.13"
>       x2="0.23" y2="0.93"
>       palindex="1" objecttype="curvedstitch"
>     />
>   </model>
> </special_stitch_models>

Author: ```suggestion
<special_stitch_models>
  <model index="0" unique_name="Rhodes Heart - over 6" name="Rhodes Heart" width="3.0" height="2.5">
    <backstitch x1="1.0" x2="2.0" y1="2.0" y2="0.0" palindex="0" objecttype="straightstitch"/>
    <backstitch x1="0.5" x2="2.5" y1="1.5" y2="0.0" palindex="0" objecttype="straightstitch"/>
    <backstitch x1="0.0" x2="3.0" y1="1.0" y2="0.5" palindex="0" objecttype="straightstitch"/>
    <backstitch x1="0.0" x2="3.0" y1="0.5" y2="1.0" palindex="0" objecttype="straightstitch"/>
    <backstitch x1="0.5" x2="2.5" y1="0.0" y2="1.5" palindex="0" objecttype="straightstitch"/>
    <backstitch x1="1.0" x2="2.0" y1="0.0" y2="2.0" palindex="0" objecttype="straightstitch"/>
    <backstitch x1="1.5" x2="1.5" y1="0.5" y2="2.5" palindex="0" objecttype="straightstitch"/>
  </model>
  <model index="1" unique_name="Lazy Daisy" name="Lazy Daisy" width="1.0" height="1.5">
    <backstitch
      x1="1.00" y1="0.00"
      x2="0.43" y2="0.26"
      x3="0.06" y3="0.66"
      x4="0.06" y4="1.03"
      x5="0.50" y5="1.06"
      x6="0.89" y6="0.66"
      x7="1.10" y7="0.10"
      palindex="0" objecttype="curvedstitch"
    />
    <backstitch
      x1="0.03" y1="1.13"
      x2="0.23" y2="0.93"
      palindex="0" objecttype="curvedstitch"
    />
  </model>
</special_stitch_models>
```

  - Currently, only unsigned number are expected to be used.

  - Stringified floating-point numbers can contain a decimal separator (period, `.`).

>     Decimal separator MUST be a period (`.`).

Reviewer: I did not do it everywhere, but some unified structure should be agreed upon

- `number`: string - Specifies both the color brand and the color number (for example, `DMC 310`).

  Brand and color number SHOULD be separated by a single space.
  Some software uses 4 spaces.

  For some reason, WinStitch adds `[+]` to the end of the `number` if the color is a blend.
  Therefore, it is RECOMMENDED to remove it to correctly parse the color number.

>   In a normalized string, the last part of the string separated by a space is the number (we expect the number to be a solid string), and everything before it is the brand.

Author: Yeah, we **separate** one value from others when writing, but then we **split** the string when parsing.

>   In a normalized string, the last part of the string separated by a space is the number (we expect the number to be a solid string), and everything before it is the brand.

Reviewer: Are you writing code or English?

>   In a normalized string, the last part of the string separated by a space is the number (we expect the number to be a solid string), and everything before it is the brand.

Author: I'm writing English which can be easily translated into the code.

>   In a normalized string, the last part of the string separated by a space is the number (we expect the number to be a solid string), and everything before it is the brand.

Reviewer: Then you should change this
Because it does not make sense in plain English

>   In a normalized string, the last part of the string separated by a space is the number (we expect the number to be a solid string), and everything before it is the brand.

Author: ```suggestion
  In a normalized string, the part following the final space indicates the color number, which is expected to be a solid string.
  Everything before that final space indicates the brand.
```

  It can be a decimal number representing a UTF-8 [code point](https://developer.mozilla.org/en-US/docs/Glossary/Code_point) or a string representing the actual character.
- `symbol_courier` _by MiniStitch (UrsaSoftware_): string - Specifies the actual symbol character (for example, `A`).
- `symbolcolor` _by XSPro Platinum (DP Software)_: rgb - Specifies the font color of the symbol.
- `bsstrands`: integer - The number of threads for stitching back stitches and other "line" stitches.
  If specified, it SHOULD be in the range of 1 to 6.
- `bscolor`: rgb - Specifies the color of a back stitch.
- `fontname` _by XSPro Platinum (DP Software)_: string - Specifies the font family (for example, `Cross Stitch Pro Platinum`) used to draw symbols of this color.
- `metalic` _by XSPro Platinum (DP Software)_: boolean - Specifies whether the color represents a metalic thread.
- `fluorescent` _by XSPro Platinum (DP Software)_: boolean - Specifies whether the color represents a fluorescent thread.
- `colorcmyk`, `bscolorcmyk`, `printcolorcmyk` _by XSPro Platinum (DP Software)_: cmyk - Specifies the color of the corresponding attribute in the CMYK color model.

>   These are important for publishers who will likely print their patterns on paper and want to achieve high repeatability thread color reproduction.

Author: I think `repeatability` and `reproduction` are the same thing. Don't mix them.
